File: lib/autoc/memory.rb
# frozen_string_literal: true


require 'singleton'
require 'autoc/std'
require 'autoc/type'
require 'autoc/module'


module AutoC


  # Standard C malloc()-based dynamic memory handler
  class Allocator

    include Singleton

    include Entity

    def initialize = dependencies << STD::STDLIB_H

    def allocate(type, count = 1, zero: false, **kws)
      zero ? "(#{type}*)calloc(#{count}, sizeof(#{type}))" : "(#{type}*)malloc((#{count})*sizeof(#{type}))"
    end

    def free(pointer) = "free(#{pointer})"

  end # Allocator


  # Boehm-Demers-Weiser garbage-collecting memory handler https://www.hboehm.info/gc/
  class BDWAllocator

    include Singleton

    include Entity

    def initialize = dependencies << SystemHeader.new('gc.h')

    def allocate(type, count = 1, atomic: false, **kws)
      atomic ? "(#{type}*)GC_malloc_atomic((#{count})*sizeof(#{type}))" : "(#{type}*)GC_malloc((#{count})*sizeof(#{type}))"
    end

    def free(pointer) = nil

  end # Allocator


end

File: src/autoc/memory.py
import functools
import autoc.core
import autoc.std as std
from autoc.core import Type
from autoc.module import Code


_allocate_code = Code(
  dependencies=(std.stdlib_h, std.assert_h, autoc.core._linkage_code),
  interface="""
    /** @private */
    AUTOC_EXTERN void* _autoc_malloc(size_t size);
    /** @private */
    AUTOC_EXTERN void* _autoc_calloc(size_t count, size_t size);
  """,
  implementation="""
    void* _autoc_malloc(size_t size) {
      void* ptr = malloc(size);
      if(!ptr && size > 0) {
        assert(0 && "malloc() returned NULL");
        abort();
      }
      return ptr;
    }
    void* _autoc_calloc(size_t count, size_t size) {
      void* ptr = calloc(count, size);
      if(!ptr && count*size > 0) {
        assert(0 && "calloc() returned NULL");
        abort();
      }
      return ptr;
    }
  """
)


# Generic C malloc()+free() memory manager
@functools.cache
class Manager(Code):
  
  def __init__(self, *args, **kws):
    super().__init__(*args, dependencies=(std.stdlib_h, _allocate_code), **kws)
    
  def allocate(self, element, count=1, zero=False, cast=None):
    if isinstance(element, Type):
      type = element
      size = f"sizeof({type})"
    else:
      type = None
      size = element
    if zero:
      code = f"_autoc_calloc({count}, {size})"
    else:
      code = f"_autoc_malloc({size})" if count == 1 else f"_autoc_malloc({count}*{size})"
    if not cast and type:
      cast = type
    return f"({cast}*){code}" if cast else code
    
  def free(self, ptr):
    return f"free({ptr})"